integration/kube_integration_test.go

```diff
@@ -1613,6 +1620,183 @@ func getContainerStatusByName(pod *v1.Pod, containerName string) *v1.ContainerSt
+	clusterConf.Auth.Preference.SetSecondFactor("off") // So we can do web login.
+
+	cluster := helpers.NewInstance(t, helpers.InstanceConfig{
+		ClusterName: clusterName,
+		HostID:      helpers.HostID,
+		NodeName:    Host,
+		Priv:        suite.priv,
+		Pub:         suite.pub,
+		Log:         suite.log,
+	})
+
+	// Setup user and role.
+	testUser := suite.me.Username
+	kubeGroups := []string{kube.TestImpersonationGroup}
+	kubeUsers := []string{testUser}
+	role, err := types.NewRole("kubemaster", types.RoleSpecV6{
+		Allow: types.RoleConditions{
+			Logins:     []string{testUser},
+			KubeGroups: kubeGroups,
+			KubeUsers:  kubeUsers,
+			KubernetesLabels: types.Labels{
+				types.Wildcard: {types.Wildcard},
+			},
+			KubernetesResources: []types.KubernetesResource{
+				{
+					Kind: types.Wildcard, Name: types.Wildcard, Namespace: types.Wildcard, Verbs: []string{types.Wildcard},
+				},
+			},
+		},
+	})
+	require.NoError(t, err)
+	cluster.AddUserWithRole(testUser, role)
+
+	err = cluster.CreateEx(t, nil, clusterConf)
+	require.NoError(t, err)
+
+	// Start the cluster.
+	err = cluster.Start()
+	require.NoError(t, err)
+	defer cluster.StopAll()
+
+	proxyAddr, err := cluster.Process.ProxyWebAddr()
+	require.NoError(t, err)
+
+	auth := cluster.Process.GetAuthServer()
+
+	userPassword := uuid.NewString()
+	require.NoError(t, auth.UpsertPassword(testUser, []byte(userPassword)))
+
+	// Login and run the tests.
+	webPack := helpers.LoginWebClient(t, proxyAddr.String(), testUser, userPassword)
+	endpoint, err := url.JoinPath("sites", "$site", "kube", kubeClusterName, "connect/ws") // :site/kube/:clusterName/connect/ws
+	require.NoError(t, err)
+
+	openWebsocketAndReadSession := func(t *testing.T, endpoint string, req web.PodExecRequest) *websocket.Conn {
+		ws, resp, err := webPack.OpenWebsocket(t, endpoint, req)
+		require.NoError(t, err)
+		require.NoError(t, resp.Body.Close())
+
+		_, data, err := ws.ReadMessage()
+		require.NoError(t, err)
+		require.Equal(t, `{"type":"create_session_response","status":"ok"}`+"\n", string(data))
+
+		execSocket := executionWebsocketReader{ws}
+
+		// First message: session metadata
+		envelope, err := execSocket.Read()
+		require.NoError(t, err)
+		var sessionMetadata sessionMetadataResponse
+		require.NoError(t, json.Unmarshal([]byte(envelope.Payload), &sessionMetadata))
+
+		return ws
+	}
+
+	findTextInReader := func(t *testing.T, reader io.ReadCloser, text string) {
+		var success atomic.Bool
+		// Make sure we don't wait forever on a read.
+		go func() {
+			time.Sleep(time.Second * 2000)
+			if !success.Load() {
+				_ = reader.Close()
+				assert.Fail(t, "Timed out waiting for command output")
+			}
+		}()
+
+		readData := make([]byte, 255)
+		accum := make([]byte, 0, 255)
+
+		for {
+			n, err := reader.Read(readData)
+			require.NoError(t, err)
+
+			accum = append(accum, readData[:n]...)
+
+			if strings.Contains(string(accum), testNamespace) {
+				success.Store(true)
+				break
+			}
+		}
+	}
```

Comment:
I don't think it will work, because it can get stuck on and individual `.Read()` and version in web tests only checks timeout between reads, so I'm inclined to keep my version.

Reply:
Right, I missed the deadline setting on a glance. So it should work, but `WStream` doesn't expose setting read deadline. And here it's more tailored to the kube exec test (like we accumulate instead of just previous string, there are many smaller chunks; we don't need remove space). I could expose `SetReadDeadline` and copy and adjust the code, but are there real benefits over this implementation?

Reply:
Changed to use read deadline as discussed offline: https://github.com/gravitational/teleport/pull/41144/commits/bc4e81b9d8a645d00a2a3c22a4c9082d09da612c